File: leetcode/dp/1692_Count_Ways_to_Distribute_Candies/codes/stirling.py
MOD = 10**9 + 7

def build_fact_and_ifact(n, mod):
    fact = [1] * (n + 1)
    for i in range(1, n + 1):
        fact[i] = fact[i - 1] * i % mod
    ifact = [1] * (n + 1)
    ifact[n] = pow(fact[n], mod - 2, mod)
    for i in range(n, 0, -1):
        ifact[i - 1] = ifact[i] * i % mod
    return fact, ifact

def nCk_mod(n, k, fact, ifact, mod):
    if k < 0 or k > n:
        return 0
    return fact[n] * ifact[k] % mod * ifact[n - k] % mod

def stirling2_mod(n, k, mod):
    if k < 0 or k > n:
        return 0
    fact, ifact = build_fact_and_ifact(k, mod)
    fact_k = fact[k]
    total = 0
    for m in range(k + 1):
        binom = nCk_mod(k, m, fact, ifact, mod)
        term = binom * pow(k - m, n, mod) % mod
        if m & 1:
            total = (total - term) % mod
        else:
            total = (total + term) % mod
    inv_fact_k = pow(fact_k, mod - 2, mod)
    return total * inv_fact_k % mod

class Solution:
    def waysToDistribute(self, n: int, k: int) -> int:
        return stirling2_mod(n, k, MOD)

File: leetcode/dp/1692_Count_Ways_to_Distribute_Candies/codes/stirling.c
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000007LL

long long modpow(long long a, long long e) {
    long long r = 1;
    while (e) {
        if (e & 1) r = r * a % MOD;
        a = a * a % MOD;
        e >>= 1;
    }
    return r;
}

void build_fact_and_ifact(int n, long long *fact, long long *ifact) {
    fact[0] = 1;
    for (int i = 1; i <= n; i++) fact[i] = fact[i - 1] * i % MOD;
    ifact[n] = modpow(fact[n], MOD - 2);
    for (int i = n; i > 0; i--) ifact[i - 1] = ifact[i] * i % MOD;
}

long long nCk_mod(int n, int k, long long *fact, long long *ifact) {
    if (k < 0 || k > n) return 0;
    return fact[n] * ifact[k] % MOD * ifact[n - k] % MOD;
}

long long stirling2_mod(int n, int k) {
    if (k < 0 || k > n) return 0;
    long long *fact = malloc((k + 1) * sizeof(long long));
    long long *ifact = malloc((k + 1) * sizeof(long long));
    build_fact_and_ifact(k, fact, ifact);
    long long fact_k = fact[k];
    long long total = 0;
    for (int m = 0; m <= k; m++) {
        long long binom = nCk_mod(k, m, fact, ifact);
        long long term = binom * modpow(k - m, n) % MOD;
        if (m & 1) total = (total - term + MOD) % MOD;
        else total = (total + term) % MOD;
    }
    long long inv_fact_k = modpow(fact_k, MOD - 2);
    free(fact);
    free(ifact);
    return total * inv_fact_k % MOD;
}

struct Solution {
    int (*waysToDistribute)(int, int);
};

int waysToDistribute_impl(int n, int k) {
    return (int)stirling2_mod(n, k);
}

struct Solution newSolution() {
    struct Solution sol;
    sol.waysToDistribute = waysToDistribute_impl;
    return sol;
}
